In [2]:
# COMP647 Assignment 02 - Data Preprocessing
# Supplement Recommendation System Based on Health and Dietary Data
# Student: Aeryun Park | Dataset: Patient Health Information 
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
pd.set_option('display.max_columns', None)  
pd.set_option('display.width', 1000)      
pd.set_option('display.max_rows', None)   
pd.set_option('display.max_colwidth', None) 

# Load the healthcare dataset containing patient information
# This dataset will be used to recommend appropriate Korure supplement categories
# based on individual health conditions and lifestyle factors
df = pd.read_csv('1167330_AeryunPark.csv')

In [3]:
# Display dataset dimensions to understand the scale of patient data
# Important for understanding the scope of our supplement recommendation model
print ("Dataset Shape:", df.shape)
print("\nColumn names:")
print(df.columns.tolist())

Dataset Shape: (1000, 20)

Column names:
['Patient_ID', 'Age', 'Gender', 'Weight_kg', 'Height_cm', 'BMI', 'Disease_Type', 'Severity', 'Physical_Activity_Level', 'Daily_Caloric_Intake', 'Cholesterol_mg/dL', 'Blood_Pressure_mmHg', 'Glucose_mg/dL', 'Dietary_Restrictions', 'Allergies', 'Preferred_Cuisine', 'Weekly_Exercise_Hours', 'Adherence_to_Diet_Plan', 'Dietary_Nutrient_Imbalance_Score', 'Diet_Recommendation']


In [4]:
# Examine the structure and first few records to understand health data format
# This helps identify key health indicators relevant for supplement classification
df.head()

Patient_ID  Age  Gender  Weight_kg  Height_cm   BMI  Disease_Type  Severity Physical_Activity_Level  Daily_Caloric_Intake  Cholesterol_mg/dL  Blood_Pressure_mmHg  Glucose_mg/dL Dietary_Restrictions Allergies Preferred_Cuisine  Weekly_Exercise_Hours  Adherence_to_Diet_Plan  Dietary_Nutrient_Imbalance_Score Diet_Recommendation
0      P0001   56    Male       58.4        160  22.8       Obesity  Moderate                Moderate                  3079              173.3                  133          116.3                  NaN   Peanuts           Mexican                    3.1                    96.6                               3.1            Balanced
1      P0002   69    Male      101.2        169  35.4      Diabetes      Mild                Moderate                  3032              199.2                  120          137.1                  NaN   Peanuts           Chinese                    4.5                    63.2                               0.6            Low_Carb
2      P0003   46  Female       63.5        173  21.2  Hypertension      Mild               Sedentary                  1737              181.0                  121          109.6                  NaN   Peanuts           Chinese                    3.8                    57.5                               4.6          Low_Sodium
3      P0004   32    Male       58.1        164  21.6           NaN      Mild                Moderate                  2657              168.2                  144          159.4                  NaN       NaN           Mexican                    4.3                    54.5                               0.4            Balanced
4      P0005   60    Male       79.5        197  20.5      Diabetes  Moderate               Sedentary                  3496              200.4                  172          182.3            Low_Sugar       NaN           Italian                    9.8                    78.2                               4.7            Low_Carb

In [5]:
df.columns.tolist()

['Patient_ID',
 'Age',
 'Gender',
 'Weight_kg',
 'Height_cm',
 'BMI',
 'Disease_Type',
 'Severity',
 'Physical_Activity_Level',
 'Daily_Caloric_Intake',
 'Cholesterol_mg/dL',
 'Blood_Pressure_mmHg',
 'Glucose_mg/dL',
 'Dietary_Restrictions',
 'Allergies',
 'Preferred_Cuisine',
 'Weekly_Exercise_Hours',
 'Adherence_to_Diet_Plan',
 'Dietary_Nutrient_Imbalance_Score',
 'Diet_Recommendation']

In [6]:
df.dtypes

Patient_ID                           object
Age                                   int64
Gender                               object
Weight_kg                           float64
Height_cm                             int64
BMI                                 float64
Disease_Type                         object
Severity                             object
Physical_Activity_Level              object
Daily_Caloric_Intake                  int64
Cholesterol_mg/dL                   float64
Blood_Pressure_mmHg                   int64
Glucose_mg/dL                       float64
Dietary_Restrictions                 object
Allergies                            object
Preferred_Cuisine                    object
Weekly_Exercise_Hours               float64
Adherence_to_Diet_Plan              float64
Dietary_Nutrient_Imbalance_Score    float64
Diet_Recommendation                  object
dtype: object

In [7]:
# Get basic statistics
df.describe()

Age    Weight_kg   Height_cm          BMI  Daily_Caloric_Intake  Cholesterol_mg/dL  Blood_Pressure_mmHg  Glucose_mg/dL  Weekly_Exercise_Hours  Adherence_to_Diet_Plan  Dietary_Nutrient_Imbalance_Score
count  1000.000000  1000.000000  1000.00000  1000.000000           1000.000000        1000.000000          1000.000000    1000.000000            1000.000000              1000.00000                       1000.000000
mean     49.857000    84.602400   174.81700    28.191600           2475.064000         199.717900           144.993000     136.867600               5.166000                74.88430                          2.469200
std      18.114267    20.088121    14.33377     8.040136            565.017032          29.080614            20.245712      37.934819               2.847995                14.82638                          1.459631
min      18.000000    50.000000   150.00000    13.000000           1500.000000         150.400000           110.000000      70.200000               0.000000                50.00000                          0.000000
25%      35.000000    66.600000   162.00000    22.075000           1984.750000         174.300000           128.000000     105.000000               2.800000                62.00000                          1.200000
50%      50.000000    85.200000   175.00000    27.450000           2470.500000         199.850000           145.000000     138.000000               5.200000                74.20000                          2.400000
75%      66.000000   102.000000   187.00000    33.425000           2937.250000         224.850000           163.000000     170.650000               7.600000                88.20000                          3.700000
max      79.000000   119.700000   199.00000    52.400000           3498.000000         249.900000           179.000000     200.000000              10.000000               100.00000                          5.000000

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 20 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   Patient_ID                        1000 non-null   object 
 1   Age                               1000 non-null   int64  
 2   Gender                            1000 non-null   object 
 3   Weight_kg                         1000 non-null   float64
 4   Height_cm                         1000 non-null   int64  
 5   BMI                               1000 non-null   float64
 6   Disease_Type                      796 non-null    object 
 7   Severity                          1000 non-null   object 
 8   Physical_Activity_Level           1000 non-null   object 
 9   Daily_Caloric_Intake              1000 non-null   int64  
 10  Cholesterol_mg/dL                 1000 non-null   float64
 11  Blood_Pressure_mmHg               1000 non-null   int64  
 12  Glucose

In [9]:
# Comprehensive missing data assessment across all health variables
# Missing health data patterns can affect supplement recommendation accuracy
df.isnull().sum()

Patient_ID                            0
Age                                   0
Gender                                0
Weight_kg                             0
Height_cm                             0
BMI                                   0
Disease_Type                        204
Severity                              0
Physical_Activity_Level               0
Daily_Caloric_Intake                  0
Cholesterol_mg/dL                     0
Blood_Pressure_mmHg                   0
Glucose_mg/dL                         0
Dietary_Restrictions                334
Allergies                           323
Preferred_Cuisine                     0
Weekly_Exercise_Hours                 0
Adherence_to_Diet_Plan                0
Dietary_Nutrient_Imbalance_Score      0
Diet_Recommendation                   0
dtype: int64

In [10]:
# Detailed analysis of categorical health columns with missing values
missing_cols = ['Disease_Type', 'Dietary_Restrictions', 'Allergies']
for col in missing_cols:
    print(f"{col} unique value:")
    print(df[col].value_counts(dropna=False))
    print("-"*30)

Disease_Type unique value:
Disease_Type
Hypertension    316
Diabetes        258
Obesity         222
NaN             204
Name: count, dtype: int64
------------------------------
Dietary_Restrictions unique value:
Dietary_Restrictions
Low_Sodium    350
NaN           334
Low_Sugar     316
Name: count, dtype: int64
------------------------------
Allergies unique value:
Allergies
Gluten     346
Peanuts    331
NaN        323
Name: count, dtype: int64
------------------------------


In [11]:
# Use domain-specific 'None' imputation for supplement recommendation context
# Missing values in health-related categorical variables likely indicate absence
# of conditions rather than unknown conditions, which is valuable for supplement targeting
df['Disease_Type'] = df['Disease_Type'].fillna('None')
df['Dietary_Restrictions'] = df['Dietary_Restrictions'].fillna('None')
df['Allergies'] = df['Allergies'].fillna('None')

# Verification that imputation was successful
df.isnull().sum()

Patient_ID                          0
Age                                 0
Gender                              0
Weight_kg                           0
Height_cm                           0
BMI                                 0
Disease_Type                        0
Severity                            0
Physical_Activity_Level             0
Daily_Caloric_Intake                0
Cholesterol_mg/dL                   0
Blood_Pressure_mmHg                 0
Glucose_mg/dL                       0
Dietary_Restrictions                0
Allergies                           0
Preferred_Cuisine                   0
Weekly_Exercise_Hours               0
Adherence_to_Diet_Plan              0
Dietary_Nutrient_Imbalance_Score    0
Diet_Recommendation                 0
dtype: int64

In [12]:
# Check Patient_ID duplicates first (most important)
patient_duplicates = df['Patient_ID'].duplicated().sum()
print(f"Patient_ID Duplicate Count: {patient_duplicates}")
print("*" * 50)

Patient_ID Duplicate Count: 0
**************************************************


In [13]:
# Optional : Quick check for suspicious duplicates
suspicious_cols = ['Weight_kg', 'Height_cm', 'BMI']
for col in suspicious_cols:
    if col in df.columns:
        dup_count = df[col].duplicated().sum()
        print(f"{col}: {dup_count} duplicates")

Weight_kg: 462 duplicates
Height_cm: 950 duplicates
BMI: 686 duplicates


In [14]:
# Check if these duplicates are normal
print("== Duplicate Analysis ==")
total_rows = len(df)
print(f"Total patients: {total_rows}")

cols_to_check = ['Weight_kg', 'Height_cm', 'BMI']
for col in cols_to_check:
    if col in df.columns:
        dup_count = df[col].duplicated().sum()
        percentage = (dup_count / total_rows) * 100
        print(f"{col}: {dup_count} duplicates ({percentage:.1f})")

== Duplicate Analysis ==
Total patients: 1000
Weight_kg: 462 duplicates (46.2)
Height_cm: 950 duplicates (95.0)
BMI: 686 duplicates (68.6)


In [15]:
# Check actual height values and their frequencies
print("== Height Value Analysis ==")
print("Most common heights:")
print(df['Height_cm'].value_counts().head(10))

print(f"\nTotal unique height: {df['Height_cm'].nunique()}")
print(f"Height range: {df['Height_cm'].min()} - {df['Height_cm'].max()}")

== Height Value Analysis ==
Most common heights:
Height_cm
170    28
177    28
190    27
184    27
168    26
198    26
183    26
160    25
157    25
196    25
Name: count, dtype: int64

Total unique height: 50
Height range: 150 - 199


In [16]:
# Identify constant features (columns with only one unique value)
# Such features provide no predictive value for supplement category classification
# However, their presence might indicate issues in health data collection
constant_features = [col for col in df.columns if df[col].nunique() == 1]
print("Constant features:", constant_features)

Constant features: []


In [17]:
# Define IQR-based outlier detection function for health metrics
# CHOICE RATIONALE: IQR method chosen for supplement recommendation because:
# 1. Health data often follows non-normal distributions
# 2. Extreme health values are meaningful for supplement targeting (e.g., severe obesity → Detox category)
# 3. Provides interpretable cutoff points for supplement category rules
def find_outliers_IQR_method(input_df, variable):
    IQR = input_df[variable].quantile(0.75) - input_df[variable].quantile(0.25)
    lower_limit = input_df[variable].quantile(0.25) - (IQR*1.5)
    upper_limit = input_df[variable].quantile(0.75) + (IQR*1.5)
    return lower_limit, upper_limit

# Systematic outlier analysis across health metrics relevant for supplement classification
# Focus on key health indicators that will inform supplement category mapping:
# - BMI, Blood Pressure, Glucose → potential Heart Health/Detox candidates
# - Cholesterol → Heart Health supplement candidates
numeriacl_cols = ['Age', 'Weight_kg', 'Height_cm', 'BMI', 'Daily_Caloric_Intake', 'Cholesterol_mg_dL', 
                  'Blood_Pressure_mmHg', 'Clucose_mg_dL', 'Weekly_Exercise_Hours']

for col in numeriacl_cols:
    if col in df.columns:
        lower, upper = find_outliers_IQR_method(df, col)
        outliers_count = len(df[(df[col] < lower) | (df[col] > upper)])
        print(f"{col}: {outliers_count} outliers (range: {lower:.1f} - {upper:.1f})")
        print("*"*50)

# Add to the end of cell 16
# Store BMI outliers for detailed analysis
lower_bmi, upper_bmi = find_outliers_IQR_method(df, 'BMI')
bmi_outliers = df[(df['BMI'] < lower_bmi) | (df['BMI'] > upper_bmi)]
print(f"\n{len(bmi_outliers)} BMI outliers stored successfully")

Age: 0 outliers (range: -11.5 - 112.5)
**************************************************
Weight_kg: 0 outliers (range: 13.5 - 155.1)
**************************************************
Height_cm: 0 outliers (range: 124.5 - 224.5)
**************************************************
BMI: 4 outliers (range: 5.1 - 50.4)
**************************************************
Daily_Caloric_Intake: 0 outliers (range: 556.0 - 4366.0)
**************************************************
Blood_Pressure_mmHg: 0 outliers (range: 75.5 - 215.5)
**************************************************
Weekly_Exercise_Hours: 0 outliers (range: -4.4 - 14.8)
**************************************************

4 BMI outliers stored successfully


In [18]:
# Show outlier details with related columns
print("\nBMI Outlier Details:")
outlier_details = bmi_outliers[['Patient_ID', 'BMI', 'Weight_kg', 'Height_cm', 'Age', 'Disease_Type']].copy()
print(outlier_details.sort_values('BMI'))


BMI Outlier Details:
    Patient_ID   BMI  Weight_kg  Height_cm  Age  Disease_Type
455      P0456  51.0      114.8        150   36  Hypertension
323      P0324  51.4      117.3        151   23  Hypertension
838      P0839  52.0      118.5        151   57      Diabetes
830      P0831  52.4      119.5        151   75       Obesity


In [19]:
# Validate BMI calculations to ensure accurate supplement category assignment
# Cross-check computed BMI against recorded BMI to prevent misclassification
print("== BMI Calculation Verification ==")
outlier_patients = ['P0456', 'P0324', 'P0839', 'P0831']

for patient_id in outlier_patients:
    patient_data = df[df['Patient_ID'] == patient_id]
    if not patient_data.empty:
        weight = patient_data['Weight_kg'].iloc[0]
        height_cm = patient_data['Height_cm'].iloc[0]
        height_m = height_cm / 100 # Convert cm to meters for BMI formula
        calculated_bmi = weight / (height_m ** 2) # Standard BMI formula: kg/m²
        recorded_bmi = patient_data['BMI'].iloc[0]
        
        print(f"Patient {patient_id}:")
        print(f"  Calculated BMI: {calculated_bmi:.1f}")
        print(f"  Recorded BMI: {recorded_bmi:.1f}")
        print(f"  Match: {abs(calculated_bmi - recorded_bmi) < 0.1}")
        print()

== BMI Calculation Verification ==
Patient P0456:
  Calculated BMI: 51.0
  Recorded BMI: 51.0
  Match: True

Patient P0324:
  Calculated BMI: 51.4
  Recorded BMI: 51.4
  Match: True

Patient P0839:
  Calculated BMI: 52.0
  Recorded BMI: 52.0
  Match: True

Patient P0831:
  Calculated BMI: 52.4
  Recorded BMI: 52.4
  Match: True



In [20]:
# DECISION: PRESERVE BMI outliers in the dataset for supplement recommendation model
# SUPPLEMENT BUSINESS JUSTIFICATION: BMI values of 51-52 represent the core target market
# for Korure Detox products. These patients are:
# 1. Most in need of weight management supplement intervention
# 2. Represent high-value customers for supplement business
# 3. Provide crucial training data for Detox category classification
# 4. Mathematically verified as accurate measurements

In [21]:
#==========================================================================================
# KORURE SUPPLEMENT CATEGORY MAPPING
#==========================================================================================

# Create target variable for supplement recommendation based on health indicators
# This mapping connects individual health conditions to appropriate Korure product categories
# following the rule-based approach outlined in the project proposal

# Korure Products

## Product Categories

### Joint Health - MP Oil, MP Powder, Relief Cream
Target: Age 65+, high BMI + older adults, high exercise hours
Focus: Joint care and mobility support

### Cardiovascular Health - Algae Oil (Plant Omega-3)
Target: High blood pressure, high cholesterol, diabetes, hypertension
Focus: Heart health and circulation

### Digestive Health - Kiwi Prebiotic + Vitamin C
Target: Moderate risk group and general health management group
Focus: Preventive health care and basic nutritional support

## Algorithm Priority (Check Order)
1. Joint Health (age/mobility issues)
2. Cardiovascular Health (heart disease risk)
3. Digestive Health (remaining patients)

In [22]:
# KORURE SUPPLEMENT CATEGORY MAPPING
# 3 Categories: Joint Health, Digestive Health

def assign_supplement_category(row):
    """
    Function to recommend appropriate Korure product category based on patient's health data
    
    Category descriptions:
    - Joint Health: Joint care products (MP Oil, MP Powder, Relief Cream)
    - Cardiovascular Health: Heart health products (Algae Oil - Plant Omega-3)
    - Digestive Health: Digestive support products (Kiwi Prebiotic + Vitamin C)
    """
    
    # Priority 1: Joint Health
    # Age-related joint degeneration, obesity-related joint stress, and overuse from excessive exercise
    if (row['Age'] >= 65 or 
        (row['BMI'] >= 35 and row['Age'] >= 55) or 
        (row['Age'] >= 45 and row['Weekly_Exercise_Hours'] >= 9)):
        return 'Joint Health'
    
    # Priority 2: Cardiovascular Health  
    # Patients with elevated cardiovascular risk markers
    if (row['Blood_Pressure_mmHg'] >= 160 or 
        row['Cholesterol_mg/dL'] >= 240 or
        row['Disease_Type'] in ['Diabetes', 'Hypertension']):
        return 'Cardiovascular Health'
    
    # Priority 3: Digestive Health
    # All remaining patients assigned to digestive health category
    # This includes patients with digestive issues, dietary restrictions, or general health maintenance needs
    return 'Digestive Health'

In [23]:
# Apply mapping function to create supplement category recommendations
print("Starting Korure product category mapping with 3 categories...")
df['Korure_Category'] = df.apply(assign_supplement_category, axis=1)

# Display mapping results
print("\n== Mapping Results ==")
category_counts = df['Korure_Category'].value_counts()
print(category_counts)

print(f"\nOut of {len(df)} total patients:")
for category, count in category_counts.items():
    percentage = (count/len(df)) * 100
    print(f"- {category}: {count} patients ({percentage:.1f}%)")

# Show sample mapping results
print("\n== Sample Mapping Results ==")
sample_data = df[['Patient_ID', 'Age', 'BMI', 'Disease_Type', 'Physical_Activity_Level', 
                  'Blood_Pressure_mmHg', 'Cholesterol_mg/dL', 'Weekly_Exercise_Hours', 'Korure_Category']].head(10)
print(sample_data.to_string(index=False))

Starting Korure product category mapping with 3 categories...

== Mapping Results ==
Korure_Category
Cardiovascular Health    475
Joint Health             346
Digestive Health         179
Name: count, dtype: int64

Out of 1000 total patients:
- Cardiovascular Health: 475 patients (47.5%)
- Joint Health: 346 patients (34.6%)
- Digestive Health: 179 patients (17.9%)

== Sample Mapping Results ==
Patient_ID  Age  BMI Disease_Type Physical_Activity_Level  Blood_Pressure_mmHg  Cholesterol_mg/dL  Weekly_Exercise_Hours       Korure_Category
     P0001   56 22.8      Obesity                Moderate                  133              173.3                    3.1      Digestive Health
     P0002   69 35.4     Diabetes                Moderate                  120              199.2                    4.5          Joint Health
     P0003   46 21.2 Hypertension               Sedentary                  121              181.0                    3.8 Cardiovascular Health
     P0004   32 21.6         No

## Actual Patient Distribution
1. Cardiovascular Health: 47.5%
2. Joint Health: 34.6%
3. Digestive Health: 17.9%

In [24]:
# Analyze key characteristics by category
print("\n== Average Health Indicators by Category ==")
category_analysis = df.groupby('Korure_Category').agg({
    'Age': 'mean',
    'BMI': 'mean',
    'Blood_Pressure_mmHg': 'mean',
    'Cholesterol_mg/dL': 'mean',
    'Glucose_mg/dL': 'mean',
    'Weekly_Exercise_Hours': 'mean'
}).round(1)

print(category_analysis.to_string())

print("\n== Final Category Distribution ==")
print("Mapping completed! New 'Korure_Category' column has been created.")
print("Categories mapped to specific Korure products:")
print("- Joint Health → MP Oil, MP Powder, Relief Cream")
print("- Cardiovascular Health → Algae Oil (Plant Omega-3)")
print("- Digestive Health → Kiwi Prebiotic + Vitamin C")
print("\nNote: This 3-category approach ensures balanced sample sizes for machine learning model training.")


== Average Health Indicators by Category ==
                        Age   BMI  Blood_Pressure_mmHg  Cholesterol_mg/dL  Glucose_mg/dL  Weekly_Exercise_Hours
Korure_Category                                                                                                
Cardiovascular Health  39.7  27.8                148.5              202.6          135.5                    5.1
Digestive Health       39.5  26.8                133.9              192.2          138.9                    4.6
Joint Health           69.1  29.5                145.9              199.7          137.7                    5.5

== Final Category Distribution ==
Mapping completed! New 'Korure_Category' column has been created.
Categories mapped to specific Korure products:
- Joint Health → MP Oil, MP Powder, Relief Cream
- Cardiovascular Health → Algae Oil (Plant Omega-3)
- Digestive Health → Kiwi Prebiotic + Vitamin C

Note: This 3-category approach ensures balanced sample sizes for machine learning model training.